src/xma/src/xmaplugin/xmaplugin.cpp

```diff
@@ -1009,58 +1061,173 @@ int32_t xma_plg_is_work_item_done(XmaSession s_handle, uint32_t timeout_ms)
         return XMA_ERROR;
     }
     priv1->using_work_item_done = true;
 
     XmaHwDevice *dev_tmp1 = priv1->device;
     if (dev_tmp1 == nullptr) {
-        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "Session XMA private pointer is NULL\n");
+        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "Session XMA private pointer is NULL");
         return XMA_ERROR;
     }
     int32_t num_execbo = priv1->num_execbo_allocated;
     if (num_execbo <= 0) {
-        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "Session XMA private: No execbo allocated\n");
+        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "Session XMA private: No execbo allocated");
         return XMA_ERROR;
     }
 
     int32_t count = 0;
-    int32_t give_up = 0;
+
+    count = priv1->kernel_complete_count;
+    if (count) {
+        priv1->kernel_complete_count--;
+        if (count > 255) {
+            xma_logmsg(XMA_WARNING_LOG, XMAPLUGIN_MOD, "CU completion count is more than 256. Application maybe slow to process CU output");
+        }
+        return XMA_SUCCESS;
+    }
     bool expected = false;
     bool desired = true;
+/*
+    while (!priv1->execbo_locked.compare_exchange_weak(expected, desired)) {
+        std::this_thread::yield();
+        expected = false;
+    }
+
+    if (xma_core::utils::check_all_execbo(s_handle) != XMA_SUCCESS) {
+        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "work_item_done->check_all_execbo. Unexpected error");
+        //Release execbo lock
+        priv1->execbo_locked = false;
+        return XMA_ERROR;
+    }
+    //Release execbo lock
+    priv1->execbo_locked = false;
 
     count = priv1->kernel_complete_count;
     if (count) {
         priv1->kernel_complete_count--;
         if (count > 255) {
-            xma_logmsg(XMA_WARNING_LOG, XMAPLUGIN_MOD, "CU completion count is more than 256. Application maybe slow to process CU output\n");
+            xma_logmsg(XMA_WARNING_LOG, XMAPLUGIN_MOD, "CU completion count is more than 256. Application maybe slow to process CU output");
         }
         return XMA_SUCCESS;
     }
-    std::this_thread::sleep_for(std::chrono::milliseconds(3));
+    if (priv1->num_cu_cmds == 0 && count == 0) {
+        xma_logmsg(XMA_WARNING_LOG, XMAPLUGIN_MOD, "Session id: %d, type: %s. There may not be any outstandng CU command to wait for\n", s_handle.session_id, xma_core::get_session_name(s_handle.session_type).c_str());
+    }
+*/
+    uint32_t iter1 = timeout_ms / 10;
+    if (iter1 < 10) {
+        iter1 = 10;
+    }
+    if (g_xma_singleton->cpu_mode == XMA_CPU_MODE1) {
+        while (iter1 > 0) {
+            std::unique_lock<std::mutex> lk(priv1->m_mutex);
+            priv1->work_item_done_1plus.wait(lk);
+            lk.unlock();
+
+            count = priv1->kernel_complete_count;
+            if (count) {
```

Comment:
Count is a int32_t, so "in theory" it could be a negative value which would cause this condition to be executed, which is not desired.   Please change count to a uint32_t value.

Reply:
Based on comments that this has many change. I will setup separate review of couple of files at a time.